File: csharp/bob/Bob.cs
using System;
using System.Text.RegularExpressions;

public static partial class Bob
{
    public static string Response(string statement)
        => statement.IsSilence() ? "Fine. Be that way!"
        : statement.IsQuestion() && statement.IsYell() ? "Calm down, I know what I'm doing!"
        : statement.IsYell() ? "Whoa, chill out!"
        : statement.IsQuestion() ? "Sure."
        : "Whatever.";

    public static bool IsSilence(this string statement) 
        => Silence().IsMatch(statement);

    public static bool IsQuestion(this string statement)
        => Question().IsMatch(statement);

    public static bool IsYell(this string statement)
        => Yell().IsMatch(statement) && !NotYell().IsMatch(statement);
    
    [GeneratedRegex("^\\s*$")]
    private static partial Regex Silence();
    [GeneratedRegex("\\?\\s*$")]
    private static partial Regex Question();
    [GeneratedRegex("[A-Z]")]
    private static partial Regex Yell();
    [GeneratedRegex("[a-z]")]
    private static partial Regex NotYell();
}

File: fsharp/bob/Bob.fs
module Bob

open System

let private someOrNone condition = if condition then Some() else None

let noLowerExists = Seq.exists Char.IsLower >> not

let someLetterExists = Seq.exists Char.IsLetter

let private allLettersUpper (text: string) =
    noLowerExists text && someLetterExists text

let (|Silence|_|) = String.IsNullOrWhiteSpace >> someOrNone

let (|Question|_|) (text: string) =
    text.TrimEnd().EndsWith('?') |> someOrNone

let (|Yell|_|) = allLettersUpper >> someOrNone

let response =
    function
    | Silence -> "Fine. Be that way!"
    | Yell & Question -> "Calm down, I know what I'm doing!"
    | Yell -> "Whoa, chill out!"
    | Question -> "Sure."
    | _ -> "Whatever."
